### PR TITLE
Add `Moment.from_ops` to more efficiently construct moments

### PR DESCRIPTION
**Pull request by maffoo**
This adds a new classmethod `cirq.Moment.from_ops` for constructing a moment from operations without calling `flatten_to_ops`. This can save a significant amount of time in workloads that construct a lot of circuits. For example, in one internal benchmark we spend almost 3% of the total runtime in the `cirq.Moment` construct (20s out of 676s total).

This also adds functions `cirq.m` and `cirq.c` as shorthands for `cirq.Moment.from_ops` and `cirq.FrozenCircuit.from_moments`, repsectively. These are similar to the `cirq.q` helper for constructing qubit types (https://github.com/quantumlib/Cirq/pull/5181) while trying to guide users to use cirq in the most performant way: avoiding the `OP_TREE` overhead associated with the default moment and circuit constructors, and preferring `FrozenCircuit` which is immutable and so can cache computed properties of the circuit. These single-letter aliases may be more controversial, but I thought I'd at least propose them. They could also be split out from this PR because I think `cirq.Moment.from_ops` is useful even without the helper functions.

### CODE REVIEW

**Comment by mpharrigan** [comment]
+1 from ops. I've hacked in my own constructor in the past for performance reasons; would be nice to have this.

The single letter things seem not very readable. Even if I grant that you can tell `cirq.m(...)` is creating a moment, I wouldn't guess that it is the more strict ops-only version. I would probably guess the opposite: shortcut is for general purpose and spelling things out gives you general behavior.

**Comment by dstrain115** [comment]
Agree with Matt.  I think the from_ops looks fine to me, but I don't think we should include the one-letter short-hand.  It's too ambiguous and having the short-hand bypass validation seems risky.

**Comment by tanujkhattar** [comment]
+1 to everything said above by Doug and Matt. 

If you want the shorthand and a way to bypass validation, I'm fine with `cirq.m(*ops, flatten=False)`; though the character savings are marginal at this point.

**Comment by mpharrigan** [comment]
```python
cc = CirqVerbatimBuilder()

circuit = cc.c([
  cc.m([cirq.H]),
  cc.m(ops),
])
```


```python
c = CirqVerbatimBuilderWithOperatorOverrides()

# use `__call__` and `__item__`
circuit = c([
  c[cirq.H, cirq.Z],
  c[ops]
])
```

**Comment by maffoo** [comment]
Thanks for the comments, everyone. I've stripped out the `cirq.c` and `cirq.m` helpers and just kept `cirq.Moment.from_ops`. PTAL.

**Comment by maffoo** [comment]
> Even if I grant that you can tell `cirq.m(...)` is creating a moment, I wouldn't guess that it is the more strict ops-only version. I would probably guess the opposite: shortcut is for general purpose and spelling things out gives you general behavior.

I agree with this in principle. Unfortunately we tried to cram a lot much flexibility into the constructors of basic data structures like `Moment` and `Circuit`, rather than putting the flexibility into helpers, so we pay a lot of extra cost whenever we construct these objects. I think it would be nice if we can build a better set of "default" APIs that are more strictly typed and avoid a lot of validation overhead, but maybe we can continue to explore this outside of cirq core like with the builder you showed.

File: cirq-core/cirq/circuits/moment.py [hunk 1]
```diff
@@ -80,17 +81,26 @@ class Moment:
             are no such operations, returns an empty Moment.
     """
 
-    def __init__(self, *contents: 'cirq.OP_TREE') -> None:
+    def __init__(self, *contents: 'cirq.OP_TREE', _flatten_contents: bool = True) -> None:
```
Comment (tanujkhattar) [inline cirq-core/cirq/circuits/moment.py:84]: Should we update the serialization to also include this flag? If not, deserialization would still be inefficient even if the serialized list of operations is flat.

**Review by tanujkhattar — APPROVED** [review]

**Comment by mpharrigan** [comment]
> Unfortunately we tried to cram a lot much flexibility into the constructors of basic data structures

yeah, I likewise agree that we messed up our constructors :/

File: cirq-core/cirq/circuits/moment.py [hunk 2]
```diff
@@ -106,6 +116,10 @@ def __init__(self, *contents: 'cirq.OP_TREE') -> None:
         self._measurement_key_objs: Optional[FrozenSet['cirq.MeasurementKey']] = None
         self._control_keys: Optional[FrozenSet['cirq.MeasurementKey']] = None
 
+    @classmethod
+    def from_ops(cls, *ops: 'cirq.Operation') -> 'cirq.Moment':
```
Comment (dabacon) [inline cirq-core/cirq/circuits/moment.py:120]: doc that this doesn't do any flattening

Comment (maffoo) [reply cirq-core/cirq/circuits/moment.py:120]: Added docstring.

**Pull request merged**